ibis/backends/flink/__init__.py

```diff
@@ -853,3 +859,142 @@ def insert(
             "is not a pandas DataFrame or is not a ibis Table."
             f"The given obj is of type {type(obj).__name__} ."
         )
+
+    def to_pyarrow(
+        self,
+        expr: ir.Expr,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        limit: int | str | None = None,
+        **kwargs: Any,
+    ) -> pa.Table:
+        from ibis.formats.pyarrow import PyArrowSchema
+
+        pyarrow_batches = self.to_pyarrow_batches(
+            expr, params=params, limit=limit, **kwargs
+        )
+
+        try:
+            pa_table = pa.Table.from_batches(pyarrow_batches)
+
+        except ValueError:
```

Comment:
`ValueError` is raised by `pa.Table.from_batches` when the input iterator `pyarrow_batches` points to an empty sequence. I could not find a way to check if the iterator points to empty sequence without advancing the iterator to the next item. However, I made the error explicit as follows:

```
        try:
            pa_table = pa.Table.from_batches(pyarrow_batches)

        except ValueError as err:
            if err.args[0] == "Must pass schema, or at least one RecordBatch":
                table_expr = expr.as_table()
                ibis_schema = table_expr.schema()
                arrow_schema = PyArrowSchema.from_ibis(ibis_schema)
                pa_table = pa.Table.from_batches(pyarrow_batches, schema=arrow_schema)
            else:
                raise ValueError(
                    "Could not convert PyArrow batches to PyArrow Table "
                    "due to an unexpected error."
                )
```

Reply:
We should try to avoid depending on the contents of exception messages. I'll push up a change that can address my concern without materializing all the data.

Reply:
I don't think the case to be handled here is super well-documented in the Python standard library docs (someone else may know better than me).

The way I handled it was using the somewhat-rare-in-practice second argument to the `next` builtin function:

```python
vals = iter([])
val1 = next(vals, None)  # evaluates to `None` because `vals` is exhausted
val2 = next(vals)  # raises StopIteration
```

In this case, I try to pull out the first batch with `next(batches, None)` and if that value is `None` then return an empty pyarrow table.

If the iterator isn't empty I recombine the first batch with the iterator using `itertools.chain`.